hashes/src/chacha20poly1305/chacha20.rs

```diff
@@ -0,0 +1,380 @@
+        let seek_offset_bytes = (seek % 64) as usize;
+        ChaCha20 { key, nonce, block_count, seek_offset_bytes }
+    }
+
+    /// Make a new instance of ChaCha20 from a block in the keystream.
+    pub fn new_from_block(key: [u8; 32], nonce: [u8; 12], block: u32) -> Self {
+        ChaCha20 { key, nonce, block_count: block, seek_offset_bytes: 0 }
+    }
+
+    /// Apply the keystream to a message.
+    pub fn apply_keystream(&mut self, to: &mut [u8]) {
+        let num_full_blocks = to.len() / CHACHA_BLOCKSIZE;
+        let mut j = 0;
+        while j < num_full_blocks {
+            let kstream =
+                keystream_at_slice(self.key, self.nonce, self.block_count, self.seek_offset_bytes);
+            for (c, k) in
+                to[j * CHACHA_BLOCKSIZE..(j + 1) * CHACHA_BLOCKSIZE].iter_mut().zip(kstream.iter())
+            {
+                *c ^= *k
+            }
+            j += 1;
+            self.block_count += 1;
+        }
+        if to.len() % 64 > 0 {
+            let kstream =
+                keystream_at_slice(self.key, self.nonce, self.block_count, self.seek_offset_bytes);
+            for (c, k) in to[j * CHACHA_BLOCKSIZE..].iter_mut().zip(kstream.iter()) {
+                *c ^= *k
+            }
+            self.block_count += 1;
+        }
+    }
+
+    /// Get the keystream block at a specified block.
+    pub fn get_keystream(&mut self, block: u32) -> [u8; 64] {
+        self.block(block);
+        keystream_at_slice(self.key, self.nonce, self.block_count, self.seek_offset_bytes)
+    }
+
+    /// Update the index of the keystream to the given byte.
+    pub fn seek(&mut self, seek: u32) {
+        self.block_count = seek / 64;
+        self.seek_offset_bytes = (seek % 64) as usize;
+    }
+
+    /// Update the index of the keystream to a block.
+    pub fn block(&mut self, block: u32) {
+        self.block_count = block;
+        self.seek_offset_bytes = 0;
+    }
+}
+
+fn quarter_round(state: &mut [u32; 16], a: usize, b: usize, c: usize, d: usize) {
+    state[a] = state[a].wrapping_add(state[b]);
+    state[d] = (state[d] ^ state[a]).rotate_left(16);
+    state[c] = state[c].wrapping_add(state[d]);
+    state[b] = (state[b] ^ state[c]).rotate_left(12);
+    state[a] = state[a].wrapping_add(state[b]);
+    state[d] = (state[d] ^ state[a]).rotate_left(8);
+    state[c] = state[c].wrapping_add(state[d]);
+    state[b] = (state[b] ^ state[c]).rotate_left(7);
+}
+
+fn double_round(state: &mut [u32; 16]) {
+    for (a, b, c, d) in CHACHA_ROUND_INDICIES {
+        quarter_round(state, a, b, c, d);
+    }
+}
+
+fn chacha_block(state: &mut [u32; 16]) {
+    let initial_state = *state;
+    for _ in 0..10 {
+        double_round(state)
+    }
+    for (modified, initial) in state.iter_mut().zip(initial_state.iter()) {
+        *modified = modified.wrapping_add(*initial)
+    }
+}
+
+fn prepare_state(key: [u8; 32], nonce: [u8; 12], count: u32) -> [u32; 16] {
+    let mut state: [u32; 16] = [0; 16];
```

Comment:
@TheBlueMatt I did a bit of experimenting here, but curious if you think I should go further. The current code with rust 1.56 (figured MSRV is a good constant) is leveraging SSE code. The newtypes I added for the secret data helped with clarity, but didn't do much in terms of performance (output is very similar, which makes sense). I started going down the path of adding a type for the quarter-round (a.k.a. the u32x4) and it starts to get pretty hairy. I noticed that the LDK code is leveraging macros to kinda "unroll" the whole round function, which was the direction I was heading, but figured I'd take a beat here since the code is already using SSE and I find the macro code pretty tough to digest.

Do you think still worth further exploring a quarter-round type? I don't have an intuition for if the current SSE output is "good enough" given the clarity it provides.